### PR TITLE
Async request metrics call support

### PR DESCRIPTION
**Pull request by shamis**
No change in erlimem is required as imem server returns the result to the process that started the session.

### CODE REVIEW

**Comment by c-bik** [comment]
Is this tested over TCP for multiple subscribers?

File: src/imem_sec.erl [hunk 1]
```diff
@@ -590,6 +591,9 @@ fetch_recs_async(SKey, Pid, Sock) ->
 fetch_recs_async(SKey, Opts, Pid, Sock) ->
     imem_statement:fetch_recs_async(SKey, Pid, Sock, Opts, true).
 
+request_metric(_SKey, Module, MetricKey, Ref, Sock) ->
+    Module:request_metric(MetricKey, Ref, Sock).
```
Comment (c-bik) [inline src/imem_sec.erl]: Should be called through `admin_exec` for `SKey` enforcement

File: src/imem_gen_metrics.erl [hunk 2]
```diff
@@ -118,6 +118,12 @@ internal_get_metric(MetricKey, ReplyFun, #state{mod=Mod, impl_state=ImplState, s
...
-build_reply_fun(ReqRef, ReplyTo) ->
-    fun(Result) -> ReplyTo ! {metric, ReqRef, Result} end.
+-spec build_reply_fun(term(), pid() | tuple()) -> fun().
+build_reply_fun(ReqRef, ReplyTo) when is_pid(ReplyTo) ->
+    fun(Result) -> ReplyTo ! {metric, ReqRef, Result} end;
+build_reply_fun(ReqRef, {Transport, Socket, Ref}) ->
+    fun(Result) ->
+        RespBin = term_to_binary({Ref, {metric, ReqRef, Result}}),
+        PayloadSize = byte_size(RespBin),
+        Transport:send(Socket, << PayloadSize:32, RespBin/binary >>)
```
Comment (c-bik) [inline src/imem_gen_metrics.erl]: Code duplication from https://github.com/K2InformaticsGmbH/imem/blob/b607135c56068adf646fe2f84a2cf5b3a03ef718/src/imem_server.erl#L209-L215
`imem_server:send_resp` should be used instead

Comment (c-bik) [inline src/imem_gen_metrics.erl]: Possibly this [`gen_server:reply`](https://github.com/K2InformaticsGmbH/erlimem/blob/master/src/erlimem_session.erl#L362) will fail for async messages

Comment (shamis) [inline src/imem_gen_metrics.erl]: @c-bik why would the `gen_server:reply` fail ?

File: src/imem_sec.erl (hunk 1 above)
Comment (shamis) [inline src/imem_sec.erl]: admin_exec cannot be used as it directly calls `apply`. In our case we need to pass the socket into the function. Should I add a function clause to match request_metric ? I have added the permission check in request_metric now. I see thats how it has been done for other functions

File: src/imem_gen_metrics.erl [hunk 3]
```diff
@@ -121,9 +121,7 @@ internal_get_metric(MetricKey, ReplyFun, #state{mod=Mod, impl_state=ImplState, s
 -spec build_reply_fun(term(), pid() | tuple()) -> fun().
 build_reply_fun(ReqRef, ReplyTo) when is_pid(ReplyTo) ->
     fun(Result) -> ReplyTo ! {metric, ReqRef, Result} end;
-build_reply_fun(ReqRef, {Transport, Socket, Ref}) ->
+build_reply_fun(ReqRef, {_Transport, _Socket, _Ref} = Sock) ->
```
Comment (c-bik) [inline src/imem_gen_metrics.erl]: `Sock` should be a opaque structure outside `imem_server`

File: src/imem_sec.erl (hunk 1 above)
Comment (c-bik) [inline src/imem_sec.erl]: Permission check is also ok. `admin_exec` can also be applied to:
```erlang
    Module:request_metric(MetricKey, Ref, Sock)
```

File: src/imem_gen_metrics.erl (hunk 2 above)
Comment (c-bik) [inline src/imem_gen_metrics.erl]: @shamis is `request_metric` single shot? or is it a subscription?

Comment (shamis) [inline src/imem_gen_metrics.erl]: @c-bik single shot. But it is requested periodically

**Comment by c-bik** [comment]
@acautin Looks good to me!

File: src/imem_server.erl [hunk 4]
```diff
@@ -187,10 +187,7 @@ mfa({Ref, Mod, Fun, Args}, Transport) ->
                end,
     ?TLog("~p MFA -> R ~n ~p:~p(~p) -> ~p~n", [Transport,Mod,Fun,NewArgs,ApplyRes]),
     ?TLog("~p MF -> R ~n ~p:~p -> ~p~n", [Transport,Mod,Fun,ApplyRes]),
-    case Ref of
-        {async, _} -> no_op;
```
Comment (acautin) [inline src/imem_server.erl]: Why was this reverted?

**Pull request merged**